compiler/lcalc/ast.ml

```diff
@@ -74,18 +89,111 @@ let make_app (e : expr Pos.marked Bindlib.box) (u : expr Pos.marked Bindlib.box
 
 let make_let_in (x : Var.t) (tau : D.typ Pos.marked) (e1 : expr Pos.marked Bindlib.box)
     (e2 : expr Pos.marked Bindlib.box) : expr Pos.marked Bindlib.box =
-  Bindlib.box_apply2
-    (fun e u -> (EApp (e, u), Pos.get_position (Bindlib.unbox e2)))
+  let pos = Pos.get_position (Bindlib.unbox e2) in
+  make_app (make_abs (Array.of_list [ x ]) e2 pos [ tau ] pos) [ e1 ] pos
+
+let ( let+ ) x f = Bindlib.box_apply f x
+
+let ( and+ ) x y = Bindlib.box_pair x y
+
+let option_enum : D.EnumName.t = D.EnumName.fresh ("eoption", Pos.no_pos)
+
+let none_constr : D.EnumConstructor.t = D.EnumConstructor.fresh ("ENone", Pos.no_pos)
+
+let some_constr : D.EnumConstructor.t = D.EnumConstructor.fresh ("ESome", Pos.no_pos)
+
+let option_enum_config : (D.EnumConstructor.t * D.typ Pos.marked) list =
+  [ (none_constr, (D.TLit D.TUnit, Pos.no_pos)); (some_constr, (D.TAny, Pos.no_pos)) ]
+
+let make_none (pos : Pos.t) =
+  (* Hack: type is not printed in to_ocaml, so I ignore it. *)
```

Comment:
You should put it as we may use option terms for other applications where the type information is needed. The type of the none case is `unit` while the type of the some case is `TAny` (should stand for the limited type polymorphism at play here with the typechecking algorithm in Dcalc).

Reply:
Ok, but how can I get the typing information from an term?

Reply:
I don't know why I resolved this. The typing information here is just `D.TAny` for the `Some` case, and `D.TUnit` for the `None` case.